protocols/identify/src/handler.rs

```diff
@@ -227,60 +239,53 @@ impl Handler {
     fn on_fully_negotiated_outbound(
         &mut self,
         FullyNegotiatedOutbound {
-            protocol: output, ..
+            protocol: output,
+            info,
         }: FullyNegotiatedOutbound<
             <Self as ConnectionHandler>::OutboundProtocol,
             <Self as ConnectionHandler>::OutboundOpenInfo,
         >,
     ) {
         match output {
-            future::Either::Left(remote_info) => {
-                self.events
-                    .push(ConnectionHandlerEvent::Custom(Event::Identified(
-                        remote_info,
-                    )));
-                self.keep_alive = KeepAlive::No;
+            future::Either::Left(substream) => {
+                let fut = Box::pin(async move {
+                    let info = protocol::recv(substream).await?;
+                    Ok(ConnectionHandlerEvent::Custom(Event::Identified(info)))
+                });
+                self.pending_events.push(fut);
```

Comment:
The difference in behaviour is that `swarm::Connection` would have raced the future returned from `OutboundUpgrade` against a timeout. See https://github.com/libp2p/rust-libp2p/blob/8f3b7e3876014aa232fa4c54cd67a145686fffc6/swarm/src/connection.rs#L461-L500.

Meaning, if successful, `output` on line 230 was the result of the future. On a timeout, `on_dial_upgrade_errror` would have been called.

With this patch, we receive the bare substream on line 242 and practically wait forever to receive a message on it.

This is what the `TimedFutures` abstraction is meant to solve: It is essentially a wrapped around `FuturesUnordered`, same as you have it here with the only addition that it will wrap each provided future in a timeout.

If you want, we wan can pair on creating this abstraction together :)

Reply:
Thanks for the explanation Thomas, implemented a `TimedFutures` based on your description and `async-bounded-queue`: c3646562c201ed9a85ea23787b286fa74a9994db. You mentioned for it to be bounded and a waker to wake when it gets empty, but you meant full right? Another question is, how can we `poll_push_ready` if `on_connection_event` is not an async function? 

>If you want, we wan can pair on creating this abstraction together :)

totally! I went ahead and bootstrapped it as you were off but we can arrange for Tomorrow, or whenever you are available.

Reply:
> Thanks for the explanation Thomas, implemented a `TimedFutures` based on your description and `async-bounded-queue`: c3646562c201ed9a85ea23787b286fa74a9994db. You mentioned for it to be bounded and a waker to wake when it gets empty, but you meant full right? 

We need both:

- A waker that wakes the `poll` task when it is empty and a new element gets pushed
- A waker that wakes the push task when it is full and a futures finishes.

Reply:
> Another question is, how can we `poll_push_ready` if `on_connection_event` is not an async function?

At the moment we can't unfortunately but as part of implementing back-pressure, we need to continue this chain eventually up to the connection task :)

For now, we can construct a no-op waker and call the function. If it is pending, we should print a warning and drop the event.

Reply:
>A waker that wakes the poll task when it is empty and a new element gets pushed

Since we are polling `FuturesUnordered` shouldn't it be enough to add to its list and poll it? Why do we need an extra `Waker` ?

Reply:
> > A waker that wakes the poll task when it is empty and a new element gets pushed
> 
> Since we are polling `FuturesUnordered` shouldn't it be enough to add to its list and poll it? Why do we need an extra `Waker` ?

I think `FuturesUnordered` will return `Poll::Ready(None)` if it has no items and not `Poll::Pending`.